In [11]:
#!pip install scikeras

#Import data

In [4]:
import pandas as pd

feature_names =  ['party','handicapped-infants', 'water-project-cost-sharing',
                    'adoption-of-the-budget-resolution', 'physician-fee-freeze',
                    'el-salvador-aid', 'religious-groups-in-schools',
                    'anti-satellite-test-ban', 'aid-to-nicaraguan-contras',
                    'mx-missle', 'immigration', 'synfuels-corporation-cutback',
                    'education-spending', 'superfund-right-to-sue', 'crime',
                    'duty-free-exports', 'export-administration-act-south-africa']

voting_data = pd.read_csv('house-votes-84.data.txt', na_values=['?'],
                          names = feature_names)
voting_data.head()

party handicapped-infants water-project-cost-sharing  \
0  republican                   n                          y   
1  republican                   n                          y   
2    democrat                 NaN                          y   
3    democrat                   n                          y   
4    democrat                   y                          y   

  adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
0                                 n                    y               y   
1                                 n                    y               y   
2                                 y                  NaN               y   
3                                 y                    n             NaN   
4                                 y                    n               y   

  religious-groups-in-schools anti-satellite-test-ban  \
0                           y                       n   
1                           y                       n   
2                           y                       n   
3                           y                       n   
4                           y                       n   

  aid-to-nicaraguan-contras mx-missle immigration  \
0                         n         n           y   
1                         n         n           n   
2                         n         n           n   
3                         n         n           n   
4                         n         n           n   

  synfuels-corporation-cutback education-spending superfund-right-to-sue  \
0                          NaN                  y                      y   
1                            n                  y                      y   
2                            y                  n                      y   
3                            y                  n                      y   
4                            y                NaN                      y   

  crime duty-free-exports export-administration-act-south-africa  
0     y                 n                                      y  
1     y                 n                                    NaN  
2     y                 n                                      n  
3     n                 n                                      y  
4     y                 y                                      y

In [5]:
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        435                 423                        387   
unique         2                   2                          2   
top     democrat                   n                          y   
freq         267                 236                        195   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                424                  424             420   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 253                  247             212   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          424                     421   
unique                           2                       2   
top                              y                       y   
freq                           272                     239   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        420       413         428   
unique                         2         2           2   
top                            y         y           y   
freq                         242       207         216   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           414                404                    410   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            264                233                    209   

       crime duty-free-exports export-administration-act-south-africa  
count    418               407                                    331  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     248               233                                    269

#Drop any missing votes
this cuts our data from 435 people to 232, there's probably a better way to do this

In [6]:
voting_data.dropna(inplace=True)
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        232                 232                        232   
unique         2                   2                          2   
top     democrat                   n                          n   
freq         124                 136                        125   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                232                  232             232   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 123                  119             128   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          232                     232   
unique                           2                       2   
top                              y                       y   
freq                           149                     124   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        232       232         232   
unique                         2         2           2   
top                            y         n           y   
freq                         119       119         128   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           232                232                    232   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            152                124                    127   

       crime duty-free-exports export-administration-act-south-africa  
count    232               232                                    232  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     149               146                                    189

Change all Y/N and Democrat/Republican to 1's and 0's

In [7]:
voting_data.replace(('y', 'n'), (1, 0), inplace=True)
voting_data.replace(('democrat', 'republican'), (1, 0), inplace=True)

In [8]:
voting_data.head()

party  handicapped-infants  water-project-cost-sharing  \
5       1                    0                           1   
8       0                    0                           1   
19      1                    1                           1   
23      1                    1                           1   
25      1                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

In [9]:
all_features = voting_data[feature_names].drop('party', axis=1).values
all_classes = voting_data['party'].values

In [12]:
import scikeras
from sklearn.model_selection import cross_val_score
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import RMSprop
from scikeras.wrappers import KerasClassifier

Define the neural network   
Dropout layers to prevent overfitting

In [14]:
def create_model():
    model=Sequential()
    model.add(Dense(32,input_dim=16,activation='relu'))
    model.add(Dropout(0.2))
    model.add(Dense(16,activation='relu'))
    model.add(Dropout(0.2))
    model.add(Dense(1,activation='sigmoid'))
    model.compile(loss='binary_crossentropy',optimizer='rmsprop',metrics=['accuracy'])
    return model

In [15]:
estimator = KerasClassifier(model=create_model,epochs=100,verbose=0)

Using cross val for accuracy measure

In [18]:
cv_scores = cross_val_score(estimator,all_features,all_classes,cv=10)
print(cv_scores.mean())

0.9396739130434784


94% isn't amazing, room for improvement.